win32ss/gdi/ntgdi/dibobj.c

```diff
@@ -129,34 +127,32 @@ CreateDIBPalette(
         {
             /* The colors are an array of RGBQUAD values */
             RGBQUAD *prgb = (RGBQUAD*)((PCHAR)pbmi + pbmi->bmiHeader.biSize);
+            RGBQUAD colors[256] = {0};
 
             // FIXME: do we need to handle PALETTEINDEX / PALETTERGB macro?
 
-            /* Loop all color indices in the DIB */
-            for (i = 0; i < cColors; i++)
+            /* Use SEH to verify we can READ prgb[] succesfully */
+            _SEH2_TRY
             {
-                /* User SEH to verify READ success */
-                Status = STATUS_SUCCESS;
-                _SEH2_TRY
-                {
-                    rgb = prgb[i];
-                }
-                _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+                for(i = 0; i < cColors; ++i)
                 {
-                    Status = _SEH2_GetExceptionCode();
-                    /* On Read Failure, put zero in Palette */
-                    PALETTE_vSetRGBColorForIndex(ppal, i, 0);
+                    colors[i] = prgb[i];
                 }
-                _SEH2_END
+            }
+            _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+            {
+              /* Do Nothing */
```

Comment:
Error handling ? Return FALSE or something ? Otherwise you're just zeroing the palette in the code below.

Reply:
Nothing is needed or wanted to be returned here. Our new statement #130 is our initializer and is as follows:
```
            RGBQUAD colors[256] = {0};
```
This effectively sets all or our 256 output RGBQUAD colors[] values to { 0, 0, 0, 0} which is what we want as our default values.
If our _SEH2_TRY succeeds in changing the value of some colors[] to something else, then we want these new values.
Now when we do the RtlCopyMemory we get the updated colors[] value copied into the 'ppal' palette.
Otherwise we want to keep the defaults and copy these into the 'ppal' and continue with our logic as a graceful fallback.

Don't you agree? Thanks, Doug Lyons

Reply:
I would prefer to return a NULL palette in this case, otherwise we are just creating a black DIB. A quick check shows that callers handle this, except here https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/dibobj.c;hb=c8ce0cc4340224bf93f4f34050deaf504d598bc3#l1460 which would need to be reworked anyway.

Reply:
If I understand what you are asking, you want something like this:
```
            /* Use SEH to verify we can READ prgb[] succesfully */
            _SEH2_TRY
            {
                RtlCopyMemory(colors, prgb, cColors * sizeof(colors[0]));
            }
            _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
            {
                Status = _SEH2_GetExceptionCode();
            }
            _SEH2_END

            if(!NT_SUCCESS(Status))
            {
                SetLastNtError(Status);
                return NULL;
            }
```

and when I tested this code, I received the following results:

![GDIProg_Bad_Exitpng](https://user-images.githubusercontent.com/32620577/123072817-7b9b3580-d3db-11eb-8b8d-838ee4140f16.png)

This is definitely NOT what Windows Server 2003 SP2 does. Do I understand your request correctly? Thanks.

Reply:
Interesting. You can leave it as it is, then.

Reply:
Great! Thanks.